### ISSUE TITLE
Audio .mp3 download very slow when using ffmpeng

### ISSUE THREAD

**Issue by Sansekai** [fent/node-ytdl-core#1186]
Example my code

```javascript
const readline = require('readline');
const ytdl = require('..');
const ffmpeg = require('fluent-ffmpeg');

let id = '7wNb0pHyGuI';

let stream = ytdl(id, {
  quality: 'highestaudio',
});

let start = Date.now();
ffmpeg(stream)
  .audioBitrate(128)
  .save(`${__dirname}/${id}.mp3`)
  .on('progress', p => {
    readline.cursorTo(process.stdout, 0);
    process.stdout.write(`${p.targetSize}kb downloaded`);
  })
  .on('end', () => {
    console.log(`\ndone, thanks - ${(Date.now() - start) / 1000}s`);
  });
```

**Comment by Brainicism** [comment]
Experiencing the same issue on many videos:
```
Downloading from https://www.youtube.com/watch?v=9bZkp7q19f0 ...
Download speed: 31.09 KB/s | 1.64% (65536/3990497)
Download speed: 33.32 KB/s | 3.28% (131072/3990497)
Download speed: 25.04 KB/s | 4.52% (180224/3990497)
...
Download speed: 23.39 KB/s | 99.77% (3981296/3990497)
Finished downloading after 135.346s @ 28.79KB/s
```
Example repro with download speed progress
```js
const ytdl = require("ytdl-core");
const fs = require("fs");

(async () => {
    const url = "https://www.youtube.com/watch?v=9bZkp7q19f0";
    console.log(`Downloading from ${url} ...`);
    const stream = ytdl(url, {
        filter: "audioonly",
        quality: "highest",
    });

    let downloadedBytesPrev = null;
    let lastUpdate = null;
    let startTime = Date.now();
    let fileTotalBytes = null;
    stream.on("progress", (_chunkLength, downloadedBytes, totalBytes) => {
        if (!downloadedBytesPrev) {
            downloadedBytesPrev = downloadedBytes;
            lastUpdate = Date.now();
            fileTotalBytes = totalBytes;
            return;
        }

        const timeSinceLastUpdate = (Date.now() - lastUpdate) / 1000;
        if (timeSinceLastUpdate > 1) {
            console.log(
                `Download speed: ${(
                    (downloadedBytes - downloadedBytesPrev) /
                    1024 /
                    ((Date.now() - lastUpdate) / 1000)
                ).toFixed(2)} KB/s | ${(
                    (100.0 * downloadedBytes) /
                    totalBytes
                ).toFixed(2)}% (${downloadedBytes}/${totalBytes})`
            );

            downloadedBytesPrev = downloadedBytes;
            lastUpdate = Date.now();
        }
    });

    stream.on("finish", () => {
        const downloadTime = (Date.now() - startTime) / 1000;
        console.log(
            `Finished downloading after ${downloadTime}s @ ${(
                fileTotalBytes /
                1024 /
                downloadTime
            ).toFixed(2)}KB/s`
        );
    });

    stream.pipe(fs.createWriteStream("./output.mp4"));
})();
```

**Comment by nextpier** [comment]
Same thing for me... This is a throttle intent by YT

**Comment by gatecrasher777** [comment]
This issue:  https://github.com/yt-dlp/yt-dlp/issues/6400

Just to repeat what I posted there:

Working on another YouTube project, today I managed to resolve all my recently experienced speed issues by adding ```&range=start-end``` to all (deciphered & n- transformed) stream urls.

Previously, you didn't need to have a range parameter in the urls, now you do otherwise the streams can be severely throttled. Having range headers in your download requests is not a substitute.

If you are using stream urls in html video or audio tags, as I do, just add ```&range=0-clen``` to the urls
with ```clen``` being the content length parameter of the same stream urls.

Then everything is fast again.

**Comment by Sansekai** [comment]
> This issue: [yt-dlp/yt-dlp#6400](https://github.com/yt-dlp/yt-dlp/issues/6400)
> 
> Just to repeat what I posted there:
> 
> Working on another YouTube project, today I managed to resolve all my recently experienced speed issues by adding `&range=start-end` to all (deciphered & n- transformed) stream urls.
> 
> Previously, you didn't need to have a range parameter in the urls, now you do otherwise the streams can be severely throttled. Having range headers in your download requests is not a substitute.
> 
> If you are using stream urls in html video or audio tags, as I do, just add `&range=0-clen` to the urls with `clen` being the content length parameter of the same stream urls.
> 
> Then everything is fast again.

can i see exmaple code?

**Comment by gatecrasher777** [comment]
@Sansekai 
a) From my downloader (simplified)

```javascript
    let start = 0;
    let end = stream.size;	
    opts.headers.Range = `bytes=${start}-${end}`;
    let videoStream = miniget(`${stream.url}&range=${start}-${end}`, opts);
```
start and end will depend on whether you are downloading the whole file or in chunks.

b) In an html video source tag:
```javascript
   let html = `<source src="${stream.url}&range=0-${stream.size}" type = "video/${stream.container}" ></source>`
```

You can get the stream size from the getinfo ```format .contentLength``` ( if given) or by parsing the url itself and using the ```clen``` parameter (if given).
The stream container is my case is either "webm" or "mp4".

**Comment by Brainicism** [comment]
Looking at @gatecrasher777 's suggestion, and looking at this ytdl codebase for the first time... this appears to work?

```diff
const getNextChunk = () => {
+ if (!rangeEnd && end >= contentLength) end = contentLength;
- if (!rangeEnd && end >= contentLength) end = 0;
  if (rangeEnd && end > rangeEnd) end = rangeEnd;
+ shouldEnd = !end || end === rangeEnd || end === contentLength;
- shouldEnd = !end || end === rangeEnd;

  requestOptions.headers = Object.assign({}, requestOptions.headers, {
    Range: `bytes=${start}-${end || ''}`,
  });

+ req = miniget(`${format.url}&range=${start}-${end}`, requestOptions);
- req = miniget(format.url, requestOptions);
```
https://github.com/fent/node-ytdl-core/blob/99e6c678eb8cf8f330479539ad08ab6a0d444322/lib/index.js#LL145C2-L154C51

**Comment by Sansekai** [comment]
> Looking at @gatecrasher777 's suggestion, and looking at this ytdl codebase for the first time... this appears to work? **Disclaimer**: This is not well tested, and I don't understand this codebase well enough to understand any edge cases
> 
> ```diff
> const getNextChunk = () => {
> + if (!rangeEnd && end >= contentLength) end = contentLength;
> - if (!rangeEnd && end >= contentLength) end = 0;
>   if (rangeEnd && end > rangeEnd) end = rangeEnd;
> + shouldEnd = !end || end === rangeEnd || end === contentLength;
> - shouldEnd = !end || end === rangeEnd;
> 
>   requestOptions.headers = Object.assign({}, requestOptions.headers, {
>     Range: `bytes=${start}-${end || ''}`,
>   });
> 
> + req = miniget(`${format.url}&range=${start}-${end  || ''}`, requestOptions);
> - req = miniget(format.url, requestOptions);
> 
> 
> + if (end && end !== rangeEnd && end !== contentLength) {
> -  if (end && end !== rangeEnd) {
> ```
> 
> https://github.com/fent/node-ytdl-core/blob/99e6c678eb8cf8f330479539ad08ab6a0d444322/lib/index.js#LL145C2-L154C51

thx, after changing the code in line 146 to 158 in https://github.com/fent/node-ytdl-core/blob/master/lib/index.js it works normally again.
```javascript
if (!rangeEnd && end >= contentLength) end = contentLength;
if (rangeEnd && end > rangeEnd) end = rangeEnd;
shouldEnd = !end || end === rangeEnd || end === contentLength;

requestOptions.headers = Object.assign({}, requestOptions.headers, {
    Range: `bytes=${start}-${end || ''}`,
});

req = miniget(`${format.url}&range=${start}-${end  || ''}`, requestOptions);
req.on('data', ondata);
req.on('end', () => {
  if (stream.destroyed) { return; }
  if (end && end !== rangeEnd && end !== contentLength) {
```

**Comment by nextpier** [comment]
@Sansekai thank you for sharing your work;  for me, after applying your patch, download is starting without throttling, but ends prematurely with a 416 error.  :-(
Tested with audio and video

**Comment by Brainicism** [comment]
Created a preliminary PR above to get things moving. To use it, should be able to `npm install` or `yarn add` `github:fent/node-ytdl-core#pull/1187/head`

@nextpier can you provide a repro?

**Comment by gatecrasher777** [comment]
Closer testing, I also found I was getting errors downloading some streams.  Whereas, viewed in html video elements, all streams worked perfectly.  So in my downloader I changed the urls to the same as I use in the video source tags..  i.e. adding just &range=0-clen, even when downloading in chunks. With ```clen``` being the ```format.contentLength.``` Now all downloads work.

Some formats like itag 22 (with both video and audio) don't have a content length and there is no &clen in the url either. Currently I calculate a contentLength with ```format.bitrate * parseInt(format.approxDurationMs) / 8000;``` which seems to work fine, but perhaps on these streams a range parameter isn't needed or isn't important.
